### PR TITLE
upgrade to use google protobuf

### PR DESCRIPTION
**Pull request by mangalaman93**
inspired from https://github.com/dgraph-io/dgraph/pull/7313

### CODE REVIEW

File: dgraph/cmd/zero/zero.go [hunk 1]
```diff
@@ -79,6 +79,7 @@ type Server struct {
 	blockCommitsOn *sync.Map
 
 	checkpointPerGroup map[uint32]uint64
+	pb.UnimplementedZeroServer
```
Comment (mangalaman93) [inline dgraph/cmd/zero/zero.go:82]: probably the same issue here

Comment (shivaji-kharse) [reply dgraph/cmd/zero/zero.go:82]: This is introduced in newer versions of protoc-gen-grpc-go. We need to embed the pb.UnimplementedZeroServer struct to ensure forward compatibility of the server. We can generate code without forward compatibility by setting the option --go-grpc_out=require_unimplemented_servers=false: on the compiler.

File: conn/raft_server.go [hunk 2]
```diff
@@ -128,6 +128,7 @@ func (p *proposals) Done(key uint64, err error) {
 type RaftServer struct {
 	m    sync.RWMutex
 	node *Node
+	pb.RaftServer
```
[the remark this answers is not in the document]
Comment (shivaji-kharse) [reply conn/raft_server.go:131]: We can avoid this by embedding the pb.UnimplementedRaftServer struct in RaftServer.

File: edgraph/server.go [hunk 3]
```diff
@@ -1179,13 +1180,13 @@ func (s *Server) State(ctx context.Context) (*api.Response, error) {
 		return nil, err
 	}
 
-	m := jsonpb.Marshaler{EmitDefaults: true}
-	var jsonState bytes.Buffer
-	if err := m.Marshal(&jsonState, ms); err != nil {
+	m := protojson.MarshalOptions{EmitUnpopulated: true}
```
[the remark this answers is not in the document]
Comment (shivaji-kharse) [reply edgraph/server.go:1183]: Both of these serve similar purposes. They include fields with zero values (such as 0, "", false) in the JSON output, ensuring that all fields defined in the proto message are present in the output JSON.

File: conn/raft_server.go [hunk 4]
```diff
@@ -128,6 +128,7 @@ func (p *proposals) Done(key uint64, err error) {
 type RaftServer struct {
 	m    sync.RWMutex
 	node *Node
+	pb.UnimplementedRaftServer
```
Comment (mangalaman93) [inline conn/raft_server.go:131]: mention here that this is for forward compatibility

File: dgraph/cmd/zero/http.go [hunk 5]
```diff
@@ -227,11 +233,17 @@ func (st *state) getState(w http.ResponseWriter, r *http.Request) {
...
 
-	m := jsonpb.Marshaler{EmitDefaults: true}
-	if err := m.Marshal(w, mstate); err != nil {
+	m := protojson.MarshalOptions{EmitUnpopulated: true}
+	buf, err := m.Marshal(mstate)
+	if err != nil {
 		x.SetStatus(w, x.ErrorNoData, err.Error())
 		return
 	}
+	_, err = w.Write(buf)
```
Comment (mangalaman93) [inline dgraph/cmd/zero/http.go:242]: move this inside if as well

File: edgraph/server.go [hunk 6]
```diff
@@ -102,7 +101,9 @@ var (
 )
 
 // Server implements protos.DgraphServer
-type Server struct{}
+type Server struct {
+	api.UnimplementedDgraphServer
```
Comment (mangalaman93) [inline edgraph/server.go:105]: mention here as well that this is for forward compatibility

File: graphql/e2e/common/admin.go [hunk 7]
```diff
@@ -485,7 +485,29 @@ func adminState(t *testing.T) {
 	}()
 	stateRes, err := io.ReadAll(resp.Body)
 	require.NoError(t, err)
-	require.NoError(t, jsonpb.Unmarshal(bytes.NewReader(stateRes), &state))
+	require.NoError(t, protojson.Unmarshal(stateRes, &state))
+
+	validateMember := func(expected *pb.Member, actual *pb.Member) {
```
Comment (mangalaman93) [inline graphql/e2e/common/admin.go:490]: why this?

File: protos/pb/sensitive.go [hunk 8]
```diff
@@ -17,8 +17,8 @@
 package pb
 
 // Sensitive implements the Stringer interface to redact its contents.
-// Use this type for sensitive info such as keys, passwords, or secrets
```
Comment (mangalaman93) [inline protos/pb/sensitive.go:20]: nit: undo this change as well

File: schema/schema_test.go [hunk 9]
```diff
@@ -0,0 +1,97 @@
+/*
+ * Copyright 2016-2024 Dgraph Labs, Inc. and Contributors
```
Comment (mangalaman93) [inline schema/schema_test.go:2]: 2024

File: x/keys.go [hunk 10]
```diff
@@ -419,7 +419,7 @@ func (p ParsedKey) ToBackupKey() *pb.BackupKey {
 	key.Attr = attr
 	key.Uid = p.Uid
 	key.StartUid = p.StartUid
-	key.Term = p.Term
+	key.Term = []byte(p.Term)
```
[the remark this answers is not in the document]
Comment (mangalaman93) [reply x/keys.go:422]: can we change the type of key.Term without changing the API?

File: dgraph/cmd/zero/http.go [hunk 11]
```diff
@@ -99,11 +99,17 @@ func (st *state) assign(w http.ResponseWriter, r *http.Request) {
 		return
 	}
 
-	m := jsonpb.Marshaler{EmitDefaults: true}
-	if err := m.Marshal(w, ids); err != nil {
+	m := protojson.MarshalOptions{EmitUnpopulated: true}
+	buf, err := m.Marshal(ids)
```
Comment (mangalaman93) [inline dgraph/cmd/zero/http.go:103]: can we use MarshalAppend here instead?

File: dgraph/cmd/zero/http.go [hunk 12]
```diff
@@ -227,11 +233,17 @@ func (st *state) getState(w http.ResponseWriter, r *http.Request) {
 		return
 	}
 
-	m := jsonpb.Marshaler{EmitDefaults: true}
-	if err := m.Marshal(w, mstate); err != nil {
+	m := protojson.MarshalOptions{EmitUnpopulated: true}
+	buf, err := m.Marshal(mstate)
```
Comment (mangalaman93) [inline dgraph/cmd/zero/http.go:237]: can we use MarshalAppend here instead?

File: graphql/admin/state.go [hunk 13]
```diff
@@ -16,25 +15,48 @@ import (
 )
 
 type membershipState struct {
```
Comment (mangalaman93) [inline graphql/admin/state.go:17]: we should avoid this change and ensure that json tags are unchanged

File: graphql/admin/state.go [hunk 14]
```diff
@@ -123,3 +147,43 @@ func convertToGraphQLResp(ms pb.MembershipState, listNs bool) membershipState {
 
 	return state
 }
+
+func mapToMember(id uint64, v *pb.Member, setId bool) Member {
```
Comment (mangalaman93) [inline graphql/admin/state.go:151]: we should not need these functions

File: worker/backup_ee.go [hunk 15]
```diff
@@ -652,16 +652,21 @@ func (tl *threadLocal) toBackupKey(key []byte) ([]byte, error) {
...
+	out := tl.alloc.Allocate(proto.Size(bk))
+	return x.MarshalToSizedBuffer(out, bk)
 }
 
 func writeKVList(list *bpb.KVList, w io.Writer) error {
-	if err := binary.Write(w, binary.LittleEndian, uint64(list.Size())); err != nil {
+	size := proto.Size(list)
+
+	// Ensure the size is non-negative and fits within uint64 bounds
+	if size < 0 {
```
Comment (mangalaman93) [inline worker/backup_ee.go:663]: we should use y.Assert here instead

File: dgraph/cmd/zero/http.go (hunk 11 above)
Comment (shivaji-kharse) [reply dgraph/cmd/zero/http.go:103]: We can use MarshalAppend, but it does not offer any benefit; it will work the same as Marshal.

Comment (mangalaman93) [reply dgraph/cmd/zero/http.go:103]: the benefit would be to be able to write to `w` directly instead of writing to buffer first.

File: go.mod [hunk 16]
```diff
@@ -10,26 +10,24 @@ require (
 	github.com/IBM/sarama v1.43.3
 	github.com/Masterminds/semver/v3 v3.3.1
 	github.com/blevesearch/bleve/v2 v2.4.3
-	github.com/dgraph-io/badger/v4 v4.4.0
-	github.com/dgraph-io/dgo/v240 v240.0.1
+	github.com/dgraph-io/badger/v4 v4.4.1-0.20241128130124-d13fdcc74a8d
```
Comment (mangalaman93) [inline go.mod:13]: let's use the latest released version of dgo and badger

File: posting/list.go [hunk 17]
```diff
@@ -1047,16 +1047,15 @@ func (l *List) ToBackupPostingList(
...
-	n, err := bl.MarshalToSizedBuffer(val)
+	val, err := x.MarshalToSizedBuffer(alloc.Allocate(proto.Size(bl)), bl)
 	if err != nil {
 		return nil, err
 	}
 
 	kv := y.NewKV(alloc)
 	kv.Key = alloc.Copy(l.key)
 	kv.Version = out.newMinTs
-	kv.Value = val[:n]
```
Comment (mangalaman93) [inline posting/list.go:1059]: why are we changing this?

File: worker/backup_ee.go [hunk 18]
```diff
@@ -652,16 +652,15 @@ func (tl *threadLocal) toBackupKey(key []byte) ([]byte, error) {
...
-	out := tl.alloc.Allocate(bk.Size())
-	n, err := bk.MarshalToSizedBuffer(out)
-	return out[:n], err
+	out := tl.alloc.Allocate(proto.Size(bk))
+	return x.MarshalToSizedBuffer(out, bk)
 }
 
 func writeKVList(list *bpb.KVList, w io.Writer) error {
-	if err := binary.Write(w, binary.LittleEndian, uint64(list.Size())); err != nil {
+	if err := binary.Write(w, binary.LittleEndian, uint64(proto.Size(list))); err != nil {
```
Comment (mangalaman93) [inline worker/backup_ee.go:660]: you can add a comment to ignore this gosec rule here
